File: OOP/1_intro/Reckless+Driver+-+Java/Reckless Driver/src/com/poash/rd/SideObjectFactory.java
package com.poash.rd;

import java.util.HashMap;

public class SideObjectFactory {
	private HashMap<Integer, Callback> callbacks = new HashMap<>() ;
	private static SideObjectFactory instance = new SideObjectFactory() ;
	private SideObjectFactory() {
		
	}
	public void RegisterCallback(int key, Callback callback) {
		callbacks.put(key, callback) ;
	}
	public SideObject CreateSideObject(int key) {
		Callback cb = callbacks.get(key) ;
		return cb.Create() ;
	}
	public static SideObjectFactory GetInstance() {
		return instance ;
	}
}

File: OOP/1_intro/Reckless+Driver+-+C++/RecklessDriver/RecklessDriver/SideObjectFactory.cpp
#include "SideObjectFactory.h"

RecklessDriver::SideObject* RecklessDriver::SideObjectFactory::CreateSideObject(int key) {
	return m_Callbacks[key]() ;
}

void RecklessDriver::SideObjectFactory::RegisterCallback(int key, Callback callback) {
	m_Callbacks[key] = callback ;
}

RecklessDriver::SideObjectFactory& RecklessDriver::SideObjectFactory::GetInstance() {
	static SideObjectFactory instance ;
	return instance ;
}
